src/internal/streamstore/streamstore.go

```diff
@@ -0,0 +1,194 @@
+
+func New(
+	kw *kopia.Wrapper,
+	tenant string,
+	service path.ServiceType,
+) *streamStore {
+	return &streamStore{kw: kw, resource: "metadata", tenant: tenant, service: service}
+}
+
+const detailsItemName = "details"
+
+// WriteBackupDetails persists a `details.Details`
+// object in the stream store
+func (ss *streamStore) WriteBackupDetails(
+	ctx context.Context,
+	backupDetails *details.Details,
+) (string, error) {
+	// construct the path of the container for the `details` item
+	p, err := path.Builder{}.
+		ToServiceCategoryMetadataPath(
+			ss.tenant,
+			ss.resource,
+			ss.service,
+			path.DetailsCategory,
+			false,
+		)
+	if err != nil {
+		return "", err
+	}
+
+	// TODO: We could use an io.Pipe here to avoid a double copy but that
+	// makes error handling a bit complicated
+	dbytes, err := json.Marshal(backupDetails)
+	if err != nil {
+		return "", errors.Wrap(err, "marshalling backup details")
+	}
+
+	dc := &streamCollection{
+		folderPath: p,
+		item: &streamItem{
+			name: detailsItemName,
+			data: dbytes,
+		},
+	}
+
+	backupStats, _, err := ss.kw.BackupCollections(ctx, nil, []data.Collection{dc}, ss.service)
+	if err != nil {
+		return "", nil
+	}
+
+	return backupStats.SnapshotID, nil
+}
+
+// ReadBackupDetails reads the specified details object
+// from the kopia repository
+func (ss *streamStore) ReadBackupDetails(
+	ctx context.Context,
+	detailsID string,
+) (*details.Details, error) {
+	// construct the path for the `details` item
+	detailsPath, err := path.Builder{}.
+		Append("details").
+		ToServiceCategoryMetadataPath(
+			ss.tenant,
+			ss.resource,
+			ss.service,
+			path.DetailsCategory,
+			true,
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	var bc stats.ByteCounter
+
+	dcs, err := ss.kw.RestoreMultipleItems(ctx, detailsID, []path.Path{detailsPath}, &bc)
+	if err != nil {
+		return nil, errors.Wrap(err, "retrieving backup details data")
+	}
+
+	// Expect only 1 data collection
+	if len(dcs) != 1 {
+		return nil, errors.Errorf("expected 1 details data collection: %d", len(dcs))
+	}
+
+	dc := dcs[0]
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, errors.New("context cancelled waiting for backup details data")
+
+		case itemData, ok := <-dc.Items():
+			if !ok {
+				return nil, errors.New("no backup details found")
+			}
+
+			var d details.Details
+
+			err := json.NewDecoder(itemData.ToReader()).Decode(&d)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to decode details data from repository")
+			}
+
+			return &d, nil
+		}
+	}
+}
+
+// DeleteBackupDetails deletes the specified details object from the kopia repository
+func (ss *streamStore) DeleteBackupDetails(
+	ctx context.Context,
+	detailsID string,
+) error {
+	err := ss.kw.DeleteSnapshot(ctx, detailsID)
+	if err != nil {
+		return errors.Wrap(err, "deleting backup details failed")
+	}
+
+	return nil
+}
+
+// streamCollection is a data.Collection used to persist
+// a single data stream
+type streamCollection struct {
```

Comment:
if you want this could be replaced by the `graph.MetadataCollection` type (same with the struct that implements `data.Stream`). We may revisit it later if we actually begin streaming data for this, but for now they both represent Collections where all data is in-memory.

If you do use the other collection, feel free to move it to a package that makes more sense

Reply:
Will do this in a follow up. There's a `statusUpdater` dependency in that package that makes moving more involved.

Reply:
ah, I forgot about that. Well, not required to consolidate code if it doesn't really fit the pattern we have already